File: C++/1st_ex/main.cpp
#include "common.h"
#include "mod_pow.h"
#include "mod_pow_log.h"
#include <iostream>

using namespace std;

int main() {
  int a, x, p;
  cin >> a >> x >> p;

  if (!is_prime(p)) {
    cout << "p must be a prime number" << endl;
    return 1;
  }

  a = a % p;

  if (a == 0)
    cout << 0 << endl;
  else {
    x = x % (p - 1);
    cout << mod_pow(a, x, p) << endl;
    cout << mod_pow_log(a, x, p) << endl;
  }
  return 0;
}

File: python/1st_ex/main.py
import common
import mod_pow
import mod_pow_log
import sys

if __name__ == '__main__':
    a = int(input())
    x = int(input())
    p = int(input())

    if not common.is_prime(p):
        print("p must be a prime number")
        sys.exit(1)

    a = a % p 

    if a == 0:
        print(0)
    else:
        x = x % (p - 1)
        print(mod_pow.mod_pow(a, x, p))
        print(mod_pow_log.mod_pow_log(a, x, p))
